[DSCP-112] Swagger spec for Student API

flyingleafe opened the pull request:
### How to review
1. Copy raw link to file `educator.yaml` or `student.yaml` from this PR
2. Go to [editor.swagger.io](https://editor.swagger.io/)
3. `File > Import URL`
4. Paste the copied link
5. Enjoy the API spec represented with a nice Swagger GUI
6. Stop enjoying and diss the spec instead

### General comments about this spec

1. Student and Educator APIs are meant to be separate and be served on two different ports, so there are two separate Swagger spec files. This is done because there are completely different authentication requirements for endpoints used by Educator and Student, and implementing them as one monolithic API really seems to be very cumbersome and error-prone. Separating the API into two completely distinct parts, where each part uses single authentication systems across its endpoints, seems to be a much cleaner solution.
2. You might find URLs too long and repetitive: e. g. why have `GET /students/{studentAddr}/grades` and `GET /grades` as two different methods? Why not instead have only `GET /grades` and a bunch of filter query parameters for it? I had the following reasoning:
    - If we'll have combined API for Educator and students, I'd like to make sure that in order to determine if a user has rights to make a request it's enough to look at method and URL path. If we'll have combined API and fully filter-based structure, a student `X` will have access to `GET /grades?student=X`, but not to `GET /grades`, which seems super counter-intuitive and ugly.
    - There are some possible query parameters which fit query string much more, such as filtering by timestamp/grade boundaries, sorting parameters, pagination, etc. I'd like to separate such parameters from parameters like `courseId`.

### What does this spec cover

- [ ] Student API
  - [x] Methods structure
  - [x] JSON schemas for API datatypes
  - [ ] Authentication
  - [ ] Response sorting
  - [x] Response filtering
  - [ ] Response pagination
  - [ ] Versioning
    - [x] (at least partially, minor versions in headers are not yet defined)

- [ ] Educator API
  - [ ] Methods structure
    - [x] (at least partially)
  - [ ] JSON schemas for API datatypes
    - [x] (at least partially)
  - [x] Authentication (no auth, serves only locally)
  - [ ] Response sorting
  - [ ] Response filtering
  - [ ] Response pagination
  - [ ] Versioning
    - [x] (at least partially, minor versions in headers are not yet defined)

flyingleafe committed:
[DSCP-112] Separate educator tech specs into DB and API specs.

flyingleafe committed:
[DSCP-112] [WIP] Swagger spec

flyingleafe commented:
I tend to think that we really should have two separate APIs for Educator integration and for Student, after thinking a bit more about authentication methods. It's more or less clear how to make a proper™ authentication in case of separate APIs:
- Educator API simply serves only localhost.
- Student API serves openly, and uses OAuth2 (or custom public-key based authorization mechanism) for student's authorization.

Given that, I think that we should deem the API which is currently described here a part of Educator API, because it has these `/student/{studentAddr}` URIs, whereas separate Student API would simply have URIs like `/courses`, `/grades`, `/submissions`, etc.

lambda76 commented:
Some thoughts, 
1. How is API versioning handled? Should we have prefix all endpoints with /api/v1/ ?
2. I prefer singular in naming, i.e /student/{studentAddr}/course/{courseId} instead of /students/{studentAddr}/courses/{courseId}
3. How should hashes and addresses be encoded? Should we encode it with BaseX something? 
4. What is the GET /grades API endpoint for? Isn't everything related to grades so this basically exports all data?
5. Some endpoints under submission section start with /students/, why are they not under student section?

Martoon-00 commented on specs/disciplina/educator/api/swagger.yaml:
```diff
@@ -0,0 +1,650 @@
+        type: "integer"
+        format: "int32"
+      timestamp:
+        $ref: "#/definitions/Time"
```
Perhaps you wished to mark it as `readOnly`.

Martoon-00 commented on specs/disciplina/educator/api/swagger.yaml:
```diff
@@ -0,0 +1,650 @@
+    authorizationUrl: "http://petstore.swagger.io/oauth/dialog"
+    flow: "implicit"
+    scopes:
+      write:pets: "modify pets in your account"
```
The truth about our relation to users has been revealed 😏

Heimdell commented on specs/disciplina/educator/api/swagger.yaml:
```diff
@@ -0,0 +1,650 @@
+        404:
+          description: "Submission with given hash not found"
+  
+  /grades:
```
Do we need such an endpoint? It is practically a database dump.

flyingleafe replied:
Not really, it only lists _grades_, and there's lots of other stuff in our database. This endpoint basically provides a way for looking at the whole private chain, like a block explorer for it.

flyingleafe committed:
[DSCP-112] Rename `swagger.yaml` to `educator.yaml`

flyingleafe committed:
[DSCP-112] Remove trailing whitespaces and redundant sample security defs.

flyingleafe committed:
[DSCP-112] Remove Student API stuff from Educator API

Removed parts of methods descriptions and a method which
only make sense if Student API and Educator API are combined.

flyingleafe committed:
[DSCP-112] Add stubs for Student API spec

flyingleafe commented:
@lambda76 , sorry for taking too long to reply.
In most my answers to your question, I'll refer to this article: https://www.vinaysahni.com/best-practices-for-a-pragmatic-restful-api , which I find quite reasonable.
1. We'll use API prefix for major version and custom headers for minor/patch version (see https://www.vinaysahni.com/best-practices-for-a-pragmatic-restful-api#versioning). This way it's easiest to introduce a new API version with breaking changes while still keeping old one functional.
2. Using plurals with object's IDs is more consistent, because you usually use plural for describing a whole resource and getting lists of objects (like `GET /students`), so it's more consistent if in order to work with a particular student we just need to append their address to resources' name (`GET /students/<addr>`, not `GET /student/<addr>`)
3. Addresses will probably use `base58` encoding (similarly to Bitcoin, Cardano and tons of others), because it's more readable and doesn't contain easy-to-mismatch pairs of characters; simple hashes and other binary data will use `base64`.
4. Not _all_ the data, there are lots of other data (courses descriptions, new assignments, ungraded submissions) which isn't yet connected with grades.
5. Look closely, they are under both `students` and `submissions` sections (because these sections is how Swagger renders _tags_. One endpoint can have many tags.)

flyingleafe committed:
[DSCP-129] Verbal spec for interaction between Student and Educator

flyingleafe commented:
Added a verbal spec for interaction between Student and Educator, which describes different options of workflow: https://github.com/DisciplinaOU/disciplina/blob/2314b0371cd18dd9072633b514eff1a321dfeb84/specs/disciplina/educator/api/spec.org

Please, read, diss and point out vague and ill-defined moments.

Heimdell reviewed (approved):
The `spec.org` part looks fine.

volhovm commented on specs/disciplina/educator/api/spec.org:
```diff
@@ -0,0 +1,120 @@
+   1) *Offline submissions*. If a course is meant to be offline (e. g. it's a
+      classic university course with course assignments done on paper) and there
+      are no actual digital _content_ of submissions, then submissions in
+      Disciplina system become effectively fictional. This means that these
```
Not exactly. They serve the role of protecting the student from the unhonest educator. Since every assignment must be signed by the student before a grade for the submission can be assigned, student can effectively stop the course at any point. Otherwise, educator can at any moment control grades for the future assignments which were not really made by the student.

volhovm reviewed (commented):
Corresponds to what we've discussed IRL.

lambda76 commented on specs/disciplina/educator/api/spec.org:
```diff
@@ -0,0 +1,120 @@
+     submissions_, they are submitted automatically as soon as final assignment
+     is assigned.
+  2) *Online final submission*. Similarly to _online course submissions_, these
+     ones should refer to some actual submission contents via _contents hash_
```
`s/submission contents via _contents hash_/submission content via _content hash_'`

lambda76 reviewed (changes requested):
These are my suggestions

flyingleafe commented on specs/disciplina/educator/api/spec.org:
```diff
@@ -0,0 +1,120 @@
+     check for updates regarding course assignments.
+     
+  These two modes can be _combined_: there can be a set of default course
+  assignments, and particular students also can receive additional ones during
```
I wrote "particular" just to emphasize that not _every_ student might receive additional assignments. I agree though, that it can be dropped.

flyingleafe commented on specs/disciplina/educator/api/spec.org:
```diff
@@ -0,0 +1,120 @@
+   _offline_ assignments, and submissions for these assignments are sent,
+   accordingly, manually or automatically.
+
+* Finishing the course
```
Why "studying _the_ course", but "completing _a_ course"?

flyingleafe committed:
[DSCP-129] Fix grammar in verbal spec

flyingleafe committed:
[DSCP-129] Fix minor bugs in verbal spec

flyingleafe committed:
[DSCP-112] Student API methods spec

flyingleafe commented:
Student API is sort of ready to diss, @lambda76 @Heimdell @Martoon-00 @volhovm  
https://github.com/DisciplinaOU/disciplina/blob/62cc4bcb8f16e6d6da3c9b1a460a40d12574977e/specs/disciplina/educator/api/student.yaml
It turned out to be quite small, but it seems that it's all we need to implement the demo app. I think we can live without a pagination for now. The only thing which left which is _necessary_ is authentication mechanism.

Also, this PR seems to be too long and large. I propose to merge it as soon as basic Student API spec is good enough, and do other things (Educator API, authentication, pagination, sorting, etc.) in separate PRs/issues.

volhovm commented on specs/disciplina/educator/api/student.yaml:
```diff
@@ -0,0 +1,414 @@
+        - in: query
+          name: enrolled
+          required: false
+          description: Is a student already enrolled in a course?
```
Sorry, what does it mean? I don't understand a semantics.

volhovm replied:
Makes sense, it should be written like this.

volhovm commented on specs/disciplina/educator/api/student.yaml:
```diff
@@ -0,0 +1,414 @@
+          description: A submission
+          required: true
+          schema:
+            $ref: '#/definitions/Submission'
```
`Submission` has "grade" field. I doubt that the student must send any grades to the educator. He should send the "ungraded submission" instead.

EDIT: "grade" field of the submission is not required, I see. The automatic example in swagger is just bad.

volhovm commented on specs/disciplina/educator/api/student.yaml:
```diff
@@ -0,0 +1,414 @@
+        '200':
+          description: successful operation
+        '403':
+          description: Cannot delete a graded submission
```
Maybe 403 should return the string with the error code so that we can process all 403 responses in a similar way?

volhovm commented on specs/disciplina/educator/api/student.yaml:
```diff
@@ -0,0 +1,414 @@
+        type: array
+        items:
+          $ref: '#/definitions/Subject'
+      is_enrolled:
```
Why this field is not required?

volhovm commented on specs/disciplina/educator/api/student.yaml:
```diff
@@ -0,0 +1,414 @@
+        format: int32
+      timestamp:
+        $ref: '#/definitions/Time'
+      has_proof:
```
What does this field mean?

volhovm replied:
Ok, so technically it's `was_published`.

flyingleafe replied:
@volhovm That would be incorrect, because the _grade_ isn't _published_. `has_proof` is way less confusing, I believe.

volhovm commented on specs/disciplina/educator/api/student.yaml:
```diff
@@ -0,0 +1,414 @@
+  Transaction:
+    type: object
+    required:
+      - hash
```
Is it a hash of block or a hash of transaction?

flyingleafe committed:
[DSCP-112] Addressing review comments for Student API spec
